src/basic/fileio.c

```diff
@@ -856,35 +857,41 @@ int read_full_file_full(
         return read_full_stream_full(f, filename, offset, size, flags, ret_contents, ret_size);
 }
 
-int executable_is_script(const char *path, char **interpreter) {
-        _cleanup_free_ char *line = NULL;
-        size_t len;
-        char *ans;
+int script_get_interpreter(const char *path, char **ret) {
+        _cleanup_fclose_ FILE *f = NULL;
         int r;
 
         assert(path);
 
-        r = read_one_line_file(path, &line);
-        if (r == -ENOBUFS) /* First line overly long? if so, then it's not a script */
-                return 0;
-        if (r < 0)
-                return r;
+        f = fopen(path, "re");
+        if (!f)
+                return -errno;
 
-        if (!startswith(line, "#!"))
-                return 0;
+        char buf[STRLEN("#!")];
+        if (fread(buf, 1, sizeof(buf), f) != sizeof(buf)) {
+                if (feof(f))
+                        return -EINVAL;
+
+                return -EIO;
+        }
+        if (!strneq(buf, "#!", sizeof(buf)))
+                return -ENOTTY;
 
```

Comment:
i think instead of using the unwieldy fread() here for two chars, i'd do something like this:

```c
static int validate_shebang(FILE *f) {
        assert(f);

        int a = getc(f);
        if (a == EOF)
                goto eof;
        if (a != (int) (unsigned char) '#')
                return -EBADMSG;
        a = getc(f);
        if (a == EOF)
                goto eof;
        if (a != (int) (unsigned char) '!')
                return -EBADMSG;
      
        return 0; 
eof:
        return feof(a) ? -EABDMSG : -errno;
}
```

or so